src/rmem.c

```diff
@@ -0,0 +1,470 @@
+#include "rmem.h"
+
+
+typedef uintptr_t ptrcmp;
+
+static size_t _AlignSize(const size_t size, const size_t align)
+{
+    return (size + (align-1)) & -align;
+}
+
+
+/************* Memory Pool *************/
+
+static void _RemoveNode(struct MemNode **const node)
+{
+    ((*node)->prev != NULL)? ((*node)->prev->next = (*node)->next) : (*node = (*node)->next);
+    ((*node)->next != NULL)? ((*node)->next->prev = (*node)->prev) : (*node = (*node)->prev);
+}
+
+struct MemPool MemPool_Create(const size_t size)
+{
+    struct MemPool mempool = {0};
+    if( size==0UL )
+        return mempool;
+    else {
+        // align the mempool size to at least the size of an alloc node.
+        mempool.stack.size = _AlignSize(size, sizeof(struct MemNode));
+        mempool.stack.mem = malloc(1 + mempool.stack.size*sizeof *mempool.stack.mem);
+        if( mempool.stack.mem==NULL ) {
+            mempool.stack.size = 0UL;
+            return mempool;
+        } else {
+            mempool.stack.base = mempool.stack.mem + mempool.stack.size;
+            return mempool;
+        }
+    }
+}
+
+struct MemPool MemPool_FromBuffer(const size_t size, void *buf)
+{
+    struct MemPool mempool = {0};
+    if( size==0UL || buf==NULL || size<=sizeof(struct MemNode) )
+        return mempool;
+    else {
+        mempool.stack.size = size;
+        mempool.stack.mem = buf;
+        mempool.stack.base = mempool.stack.mem + mempool.stack.size;
+        return mempool;
+    }
+}
+
+void MemPool_Destroy(struct MemPool *const mempool)
+{
+    if( mempool==NULL || mempool->stack.mem==NULL )
+        return;
+    else {
+        free(mempool->stack.mem);
+        *mempool = (struct MemPool){0};
+    }
+}
+
+void *MemPool_Alloc(struct MemPool *const mempool, const size_t size)
+{
+    if( mempool==NULL || size==0UL || size > mempool->stack.size )
+        return NULL;
+    else {
+        struct MemNode *new_mem = NULL;
+        const size_t ALLOC_SIZE = size + sizeof *new_mem;
+        if( mempool->freeList.head != NULL ) {
+            const size_t MEM_SPLIT_THRESHOLD = sizeof(intptr_t);
+            // if the freelist is valid, let's allocate FROM the freelist then!
+            for( struct MemNode **inode = &mempool->freeList.head; *inode != NULL; inode = &(*inode)->next ) {
+                if( (*inode)->size < ALLOC_SIZE )
+                    continue;
+                else if( (*inode)->size <= ALLOC_SIZE + MEM_SPLIT_THRESHOLD ) {
+                    // close in size - reduce fragmentation by not splitting.
+                    new_mem = *inode;
+                    _RemoveNode(inode);
+                    mempool->freeList.len--;
+                    new_mem->next = new_mem->prev = NULL;
+                    break;
```

Comment:
Useless break.

Reply:
it is not useless.

Reply:
Ah, sorry my bad. Same comment about `else` before applies to `else`s here.